bpf/lib/proxy.h

```diff
@@ -10,6 +10,149 @@
 #error "Proxy redirection is only supported from skb context"
 #endif
 
+#ifdef ENABLE_TPROXY
+static __always_inline int
+assign_socket_tcp(struct __ctx_buff *ctx,
+		  struct bpf_sock_tuple *tuple, __u32 len, bool established)
+{
+	int result = DROP_PROXY_LOOKUP_FAILED;
+	struct bpf_sock *sk = NULL;
+	__u16 dbg_ctx;
+
+	sk = skc_lookup_tcp(ctx, tuple, len, BPF_F_CURRENT_NETNS, 0);
+	if (!sk)
+		goto out;
+
+	if (established && sk->state == BPF_TCP_TIME_WAIT)
+		goto release;
+	if (established && sk->state == BPF_TCP_LISTEN)
+		goto release;
+
+	dbg_ctx = sk->family << 16 | ctx->protocol;
+	result = sk_assign(ctx, sk, 0);
+	cilium_dbg(ctx, DBG_SK_ASSIGN, -result, dbg_ctx);
+	if (result == 0)
+		result = CTX_ACT_OK;
+	else
+		result = DROP_PROXY_SET_FAILED;
+release:
+	sk_release(sk);
+out:
+	return result;
+}
+
+static __always_inline int
+assign_socket_udp(struct __ctx_buff *ctx,
+		  struct bpf_sock_tuple *tuple, __u32 len,
+		  bool established __maybe_unused)
+{
+	int result = DROP_PROXY_LOOKUP_FAILED;
+	struct bpf_sock *sk = NULL;
+	__u16 dbg_ctx;
+
+	sk = sk_lookup_udp(ctx, tuple, len, BPF_F_CURRENT_NETNS, 0);
+	if (!sk)
+		goto out;
+
+	dbg_ctx = sk->family << 16 | ctx->protocol;
+	result = sk_assign(ctx, sk, 0);
+	cilium_dbg(ctx, DBG_SK_ASSIGN, -result, dbg_ctx);
+	if (result == 0)
+		result = CTX_ACT_OK;
+	else
+		result = DROP_PROXY_SET_FAILED;
+	sk_release(sk);
+out:
+	return result;
+}
+
+static __always_inline int
+assign_socket(struct __ctx_buff *ctx,
+	      struct bpf_sock_tuple *tuple, __u32 len,
+	      __u8 nexthdr, bool established)
+{
+	/* Workaround: While the below functions are nearly identical in C
+	 * implementation, the 'struct bpf_sock *' has a different verifier
+	 * pointer type, which means we can't fold these implementations
+	 * together.
+	 */
+	switch (nexthdr) {
+	case IPPROTO_TCP:
+		return assign_socket_tcp(ctx, tuple, len, established);
+	case IPPROTO_UDP:
+		return assign_socket_udp(ctx, tuple, len, established);
+	}
+	return DROP_PROXY_UNKNOWN_PROTO;
+}
+
+/**
+ * combine_ports joins the specified ports in a manner consistent with
+ * pkg/monitor/dataapth_debug.go to report the ports ino monitor messages.
+ */
+static __always_inline __u32
+combine_ports(__u16 dport, __u16 sport)
+{
+	return (bpf_ntohs(dport) << 16) | bpf_ntohs(sport);
+}
+
+/**
+ * ctx_redirect_to_proxy_ingress4 / ctx_redirect_to_proxy_ingress6
+ * @ctx				pointer to program context
+ * @tuple			pointer to *scratch buffer* with packet tuple
+ * @proxy_port			port to redirect traffic towards
+ *
+ * Prefetch the proxy socket and associate with the ctx. Must be run on tc
+ * ingress. Will modify 'tuple'!
+ */
+#define CTX_REDIRECT_FN(NAME, CT_TUPLE_TYPE, SK_FIELD,				\
+			DBG_LOOKUP_CODE, DADDR_DBG, SADDR_DBG)			\
+static __always_inline int							\
+NAME(struct __ctx_buff *ctx, CT_TUPLE_TYPE * ct_tuple, __be16 proxy_port)	\
```

Comment:
I see this is a macro definition, but the docstring doesn't match. It's instead the docstring of a usage of the macro (further down below). I think it would make sense to move the `ctx_redirect_to_proxy_ingress4 / 6` to where it's actually used and further document the macro. It's sort of confusing at first glance. If this was intentional though, I'm curious why :)

Also the spacing in the docstrings for the args are a bit messed up.

Reply:
I wasn't planning on documenting the macro, so I figured it'd be more useful to provide examples and actual types of the declared functions for reference here to make it easier to reason about the logic defined by the macro. Perhaps if I moved the docstring inside the macro definition it would be more clear that it applies to the functions declared by the macro.

Also, I specifically wanted to include the function names explicitly here so that it's easier to grep and find this usage and know that this macro is defining the function to provide a hint to readers that "yes this is the piece you're looking for".

The spacing is not messed up if you open your editor using the standard settings for kernel dev (tabs=8wide), this is a known conflict with git (apparently including github) rendering of the lines.

Reply:
I've queued up a change to move the docstrings inside the macros.